### PR TITLE
MINOR: Add documentation about harness runner

### PR DESCRIPTION
**Pull request by bbejeck**
Covers #403

### CODE REVIEW

**Comment by bbejeck** [comment]
ping @ybyzek and @colinhicks for review

**Comment by bbejeck** [comment]
I'll make a pass over the grammar soon.

File: README.md [hunk 1]
```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
...
+          ms: 30000
+          render:
+            skip: true
+```
+
+
+* Title - each section starts with a `title` element and as the label suggests, the text provided here is the text used the label the step for the tutorial user and the output to the console by the harness runner.  The `title` section contains one element - `content`
+
+* content - the `content` section (a YAML `dictionary`) contains an arbitrary sized list of YAML dictionaries named `action`. An `action` key creates an anonymous step i.e. not specified in the output of the test runner.  For output in the test runner, you can provide a `-name` key followed by some text for console output.
+* action - `action` keys drive the behavior of the harness. An action key can be one of these values
```
Comment (colinhicks) [inline README.md:330]: I noticed this section does not discuss `action: docker_ksql_cli_session`. This type probably deserves its own description below.

**Comment by colinhicks** [comment]
Thanks for adding this guide, @bbejeck. The content seems good at a high level. I will hold off on specific suggestions until after you make the proof-reading pass you mentioned.

File: README.md (hunk 1 above)
Comment (bbejeck) [reply README.md:330]: ack, will do!

**Comment by ybyzek** [comment]
@bbejeck thanks for the PR!  I copied @vdesabou and @framiere to review.

@vdesabou -- could you kindly review this PR with the intention to get it to the state that it meets the need to educate users how to test each Kafka Tutorial in their own environment (and thus remove the need for duplicating this in your repo).

File: README.md [hunk 2]
```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
 
 Lastly, create a Makefile in the `code` directory to invoke the harness runner and check any outputs that it produces. Then modify the `.semaphore/semaphore.yml` file to invoke that Makefile. This will make sure your tutorial gets checked by the CI system.
 
+## Working with the harness file
```
Comment (ybyzek) [inline README.md:240]:
```suggestion
## Testing Locally
```

File: README.md [hunk 3]
```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
...
+
+Given the test harness is the `heart` of a tutorial, it will be helpful to describe in detail how to work with a `kafka|ksql|kstreams.yml` file.  It should be noted the harness file is in the [YAML file formt](https://en.wikipedia.org/wiki/YAML), so formatting properly is essential.  The harness files generates the structure of the rendered tutorial and also validates any output of tutorial steps against expected values.
```
Comment (ybyzek) [inline README.md:242]: @bbejeck can we had a Table of Contents at the top of the README?

File: README.md [hunk 4]
```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
...
+
+This section is not meant to be an exhaustive decription of the harness file.  New tutorial authors should not need to create a harness file from scratch, using either the `tools/gen_project.sh` or  `tools/clone.sh` script will provide a usable harness file.  This section should provide enough guidance to add, update, or remove sections as needed.
+
+#### TLDR;
+
+To run a tutorial programatically do the following stepes. Note these instructions assume you've already checked out the kafka-tutorials repo).
```
Comment (ybyzek) [inline README.md:248]: Why assume?  Can we add a step:

```
git clone https://github.com/confluentinc/kafka-tutorials.git
cd kafka-tutorials
```

File: README.md [hunk 5]
```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
...
+
+1. cd into `_includes/tutorials/<tutorial name>/<type>/code` directory where type is one of `ksql | kstreams | kafka`.
+2. run the `make` command.
+
+The `Makefile` will delete and re-create the `outputs` directory used contain files with output from various steps used to verify the tutorial steps.
+
+Here's an example of the contents of a `Makefile`:
```
Comment (ybyzek) [inline README.md:255]: FYC: hyperlink to an existing Makefile

File: README.md [hunk 6]
```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
...
+
+Given the test harness is the `heart` of a tutorial, it will be helpful to describe in detail how to work with a `kafka|ksql|kstreams.yml` file.  It should be noted the harness file is in the [YAML file formt](https://en.wikipedia.org/wiki/YAML), so formatting properly is essential.  The harness files generates the structure of the rendered tutorial and also validates any output of tutorial steps against expected values.
+
+This section is not meant to be an exhaustive decription of the harness file.  New tutorial authors should not need to create a harness file from scratch, using either the `tools/gen_project.sh` or  `tools/clone.sh` script will provide a usable harness file.  This section should provide enough guidance to add, update, or remove sections as needed.
+
+#### TLDR;
+
+To run a tutorial programatically do the following stepes. Note these instructions assume you've already checked out the kafka-tutorials repo).
+
+1. cd into `_includes/tutorials/<tutorial name>/<type>/code` directory where type is one of `ksql | kstreams | kafka`.
```
Comment (ybyzek) [inline README.md:250]: Add example.

```For example, if you want to run the test the Kafka Tutorial at https://kafka-tutorials.confluent.io/create-stateful-aggregation-count/kstreams.html, do this:

cd _includes/tutorials/aggregating-count/kstreams/code
```

File: README.md [hunk 7]
```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
...
+2. run the `make` command.
```
Comment (ybyzek) [inline README.md:251]: I tried to run this myself and got:

```
“The ‘harness-runner==0.0.1’ distribution was not found and is required by the application”
```

1. How can I resolve this?
2. Consider whether this PR should add the explanation to resolve this error message

File: README.md [hunk 8]
```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
...
+
+* dev - the setup and teaching portion of the tutorial (required)
+* test - test setup and execution of tests, if any (optional)
+* prod - steps to build and deploy a docker image of the tutorial code (optional)
+
+In some cases, having a test and or prod section doesn't make sense, so those sections can be left out of the harness file.  The Apache Kafka console producer and consumer tutorials are a good example of tutorials that don't need a test or prod section.
+
+The `dev`, `test`, and `prod` sections contain a top-level element `steps`.  The `steps` contains any number of well, steps for the user to walk through.  Addtionally the `harness_runner` script follows the same steps for executing the tutorial automatically during builds.  The steps in any section are structured in the same way, so we'll only discuss the make-up of a single section.
+
+For reference here is an example section
```
Comment (ybyzek) [inline README.md:299]: Hyperlink to example file?

File: README.md [hunk 9]
```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
 
 Lastly, create a Makefile in the `code` directory to invoke the harness runner and check any outputs that it produces. Then modify the `.semaphore/semaphore.yml` file to invoke that Makefile. This will make sure your tutorial gets checked by the CI system.
 
+## Working with the harness file
+
+Given the test harness is the `heart` of a tutorial, it will be helpful to describe in detail how to work with a `kafka|ksql|kstreams.yml` file.  It should be noted the harness file is in the [YAML file formt](https://en.wikipedia.org/wiki/YAML), so formatting properly is essential.  The harness files generates the structure of the rendered tutorial and also validates any output of tutorial steps against expected values.
+
+This section is not meant to be an exhaustive decription of the harness file.  New tutorial authors should not need to create a harness file from scratch, using either the `tools/gen_project.sh` or  `tools/clone.sh` script will provide a usable harness file.  This section should provide enough guidance to add, update, or remove sections as needed.
+
+#### TLDR;
```
Comment (ybyzek) [inline README.md:246]:
```suggestion
#### TL;DR
```

File: README.md [hunk 10]
```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
...
+
+#### 1. Structure
+
+Three top-level sections make up the harness file:
+
+* dev - the setup and teaching portion of the tutorial (required)
+* test - test setup and execution of tests, if any (optional)
+* prod - steps to build and deploy a docker image of the tutorial code (optional)
+
+In some cases, having a test and or prod section doesn't make sense, so those sections can be left out of the harness file.  The Apache Kafka console producer and consumer tutorials are a good example of tutorials that don't need a test or prod section.
```
Comment (ybyzek) [inline README.md:295]: Can you add a hyperlink to those tutorials?

File: README.md [hunk 11]
```diff
@@ -237,6 +237,180 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
...
+* `execute` The `execute` step is a synchronous execution step during the tutorial for the user as well as the harness runner
+
+    ```yml
+     - action: execute
+          file: tutorial-steps/dev/init.sh
+          render:
+            file: tutorials/console-consumer-primitive-keys-values/kafka/markup/dev/init.adoc
+    ```
+
+    Somtimes the `execute` step is an _internal_ step only for the harness runner. Internal steps ignore the `render:` using a `skip: true` entry
```
Comment (ybyzek) [inline README.md:367]:
```suggestion
    Sometimes the `execute` step is an _internal_ step only for the harness runner. Internal steps ignore the `render:` using a `skip: true` entry
```

File: README.md (hunk 7 above)
Comment (vdesabou) [reply README.md:251]: It would be good to have a dedicated `prerequisites` section just for being able to test locally. Right now there is ruby, bundle, etc.. required, but I'm not sure this is needed just to run the harness-runner ? 
Ideally a docker version of the harness-runner would be great so that users don't have to install python3 and pip3

Comment (ybyzek) [reply README.md:251]: @bbejeck as a reference example: 

https://github.com/confluentinc/examples/blob/5.5.0-post/clients/cloud/python/README.md#example-3-run-all-the-above-in-docker
https://github.com/confluentinc/examples/blob/5.5.0-post/clients/cloud/python/requirements.txt
https://github.com/confluentinc/examples/blob/5.5.0-post/clients/cloud/python/Dockerfile

I think you were thinking of addressing this in a separate PR.  But if you help me debug the harness runner error, I'm happy to build the Docker stuff.

File: README.md (hunk 3 above)
Comment (bbejeck) [reply README.md:275]: @ybyzek I took a stab at this, check it out and let me know WYT.

File: README.md (hunk 4 above)
Comment (bbejeck) [reply README.md:281]: ack

File: README.md (hunk 6 above)
Comment (bbejeck) [reply README.md:283]: ack

File: README.md (hunk 5 above)
Comment (bbejeck) [reply README.md:288]: ack

File: README.md (hunk 8 above)
Comment (bbejeck) [reply README.md:332]: ack

File: README.md (hunk 10 above)
Comment (bbejeck) [reply README.md:328]: ack

File: README.md [hunk 12]
```diff
@@ -237,6 +268,210 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
...
+
+* content - the `content` section (a YAML `dictionary`) contains an arbitrary sized list of YAML dictionaries named `action`. An `action` key creates an anonymous step i.e. not specified in the output of the test runner.  For output in the test runner, you can provide a `-name` key followed by some text for console output.
+* action - `action` keys drive the behavior of the harness. An action key can be one of these values
+    * `action: make_file` - an action prompting the user to create a file for the tutorial.
+    * `action: execute` - a synchronous action step
+    * `action: execute_async` - an asynchronous step, this indicates a step the user will keep running for some portion of the tutorial.
+    * `action: sleep` - pause the test runner for an amount of time specified by the `ms` key. You use `sleep` key to allow some asnyc action to complete
+        * `ms: NNN` - the time in milliseconds you want the test harness to pause execution.  You only use `ms` after an `action: sleep` entry.
+    * `docker_ksql_cli_session` - an action starting a ksqlDB CLI session for working through a tutorial
```
Comment (bbejeck) [inline README.md:372]: @colinhicks I've added a section for `docker_ksql_cli_session`

File: README.md (hunk 3 above)
Comment (ybyzek) [reply README.md:273]: @bbejeck Awesome!  For your consideration: decrease it 2 levels instead of 3 levels -- makes it easier for future maintenance so you're not chasing down changes at the 3rd level back to the TOC.  But if you prefer 3, keep it ;). Looks great!

File: README.md (hunk 7 above)
Comment (ybyzek) [reply README.md:288]: Dockerfile:

```
FROM python:3.7-slim
RUN pip3 install pyyaml
```

Build and execute:

```
docker build -t runner . ; docker run -v ${PWD}/harness_runner:/harness_runner/ -it --rm runner bash -c 'cd /harness_runner/ && pip3 install -e .'
```

And now make works:

```
(cd _includes/tutorials/transforming/kstreams/code/ && make)
```

File: README.md [hunk 13]
```diff
@@ -237,6 +268,210 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
...
+Given the test harness is the `heart` of a tutorial, it will be helpful to describe in detail how to work with a `kafka|ksql|kstreams.yml` file.  It should be noted the harness file is in the [YAML file formt](https://en.wikipedia.org/wiki/YAML), so formatting properly is essential.  The harness files generates the structure of the rendered tutorial and also validates any output of tutorial steps against expected values.
+
+This section is not meant to be an exhaustive decription of the harness file.  New tutorial authors should not need to create a harness file from scratch, using either the `tools/gen_project.sh` or  `tools/clone.sh` script will provide a usable harness file.  This section should provide enough guidance to add, update, or remove sections as needed.
+
+#### TL;DR
+
+To run a tutorial programatically do the following stepes. Note to follow these instructions you'll need to checkout the kafka-tutorials repo:
+
+1. git clone https://github.com/confluentinc/kafka-tutorials.git
+2. cd kafka-tutorials
```
Comment (ybyzek) [inline README.md:282]: Should we add a 3rd step to setup environment per the instructions at https://github.com/confluentinc/kafka-tutorials#setup ?  If so, that might be where we provide the Dockerfile solution?

Comment (ybyzek) [reply README.md:282]: Or is it the case that running the harness_runner does NOT require all those other dependencies?

File: README.md [hunk 14]
```diff
@@ -237,6 +268,210 @@ You can do the same for Kafka Streams and Kafka, by using the `kstreams` and `ka
 
 Lastly, create a Makefile in the `code` directory to invoke the harness runner and check any outputs that it produces. Then modify the `.semaphore/semaphore.yml` file to invoke that Makefile. This will make sure your tutorial gets checked by the CI system.
 
+## Testing Locally
+
+Given the test harness is the `heart` of a tutorial, it will be helpful to describe in detail how to work with a `kafka|ksql|kstreams.yml` file.  It should be noted the harness file is in the [YAML file formt](https://en.wikipedia.org/wiki/YAML), so formatting properly is essential.  The harness files generates the structure of the rendered tutorial and also validates any output of tutorial steps against expected values.
```
Comment (vdesabou) [inline README.md:273]: Typo: `YAML file formt`

File: README.md (hunk 13 above)
Comment (ybyzek) [reply README.md:282]: Just wanted to document a suggestion from @vdesabou here -- this differentiation would address the confusion I had and help other users as well.

> It would be great to have a clarification in prerequisite section to specify that dependencies are only to build the site and not required to test locally.

File: README.md (hunk 3 above)
Comment (bbejeck) [reply README.md:273]: @ybyzek great idea, will do

File: README.md (hunk 14 above)
Comment (bbejeck) [reply README.md:253]: Good catch!

**Comment by bbejeck** [comment]
I've updated this per comments

\cc @ybyzek and @vdesabou

**Comment by bbejeck** [comment]
@colinhicks I'm going to merge this, but I'll make any changes you want in a follow-up PR.

**Comment by bbejeck** [comment]
No code changes are in this PR, merging now.
